server/src/main/java/org/elasticsearch/search/slice/DocIdSliceQuery.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.slice;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.ConstantScoreScorer;
+import org.apache.lucene.search.ConstantScoreWeight;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.TwoPhaseIterator;
+import org.apache.lucene.search.Weight;
+
+import java.io.IOException;
+
+/**
+ * A {@link SliceQuery} that partitions documents based on their Lucene ID.
+ *
+ * NOTE: Because the query relies on Lucene document IDs, it is not stable across
+ * readers. It's intended for scenarios where the reader doesn't change, like in
+ * a point-in-time search.
+ */
+public final class DocIdSliceQuery extends SliceQuery {
```

Comment:
Some notes:
* In principle we could add support for using this query as a lead iterator. This didn't seem so useful to me since in the common case, we shouldn't be slicing within a shard in a super fine-grained way?
* I considered slicing based on contiguous ranges of document IDs. This had the advantage that slices could skip certain segments completely. However I was worried about the case where sort order was heavily correlated with doc ID order, which would result in a big imbalance across slices (maybe some would even be empty).

Reply:
I like the idea of contiguous ranges. Slicing is useful for a full scan where fetching the data is most of the time the costly operation. Using sequential doc ids per slice would ensure that a match all query with slices can get the fetch optimization all the time (read stored_fields per block).

Reply:
I thought more and am not sure my concern even made sense. Slicing is used when scanning over all documents, not to find top hits, and it's fine if some slices skip over all the early results.

I pushed a simple approach where we split the documents into ranges with roughly equivalent size.